Jenkinsfile

```diff
@@ -1251,7 +1251,7 @@ pipeline {
                 stage('Functional on Leap 15') {
                     when {
                         beforeAgent true
-                        expression { ! skip_ftest('leap15') }
+                        expression { ! skip_ftest('leap15', env.BRANCH_NAME == 'master') }
```

Comment:
Why this? And even if so, the logic belongs in the function, not in the `pipeline {}` block.  No unnecessary logic in the `pipeline {}` block, even if it seems to make more sense.  Our goal in the `pipeline {}` is the least amount of code/logic, not what makes the most sense.

Reply:
It takes a lot of time to run, and is very unlikely to fail, it's basically an integration test and could easily be pushed back to weekly testing.

Reply:
> It takes a lot of time to run,

A lot?  Relative to what?  It shouldn't take materially longer than the EL VM test run.  And it's running on resources that we have plenty of and can stand more of up quite easily.

While I agree that we should see how we can utilise resources that are scarce better, I don't think we should be eliminating testing on resources that are plentiful.

Reply:
Good point, this runs on vm9 doesn't it?  I know it was a pinch point recently but you've solved that with zypper changes haven't you?

It's still a duplicated test which is unlikely to fail with anything other than integration issues.

Reply:
> Good point, this runs on vm9 doesn't it?

It does.  Which, modulo bugs, etc. (such as the Zypper issue)  is not a constrained resource and is easily unconstrained if it becomes such.

> I know it was a pinch point recently but you've solved that with zypper changes haven't you?

Correct.
 
> It's still a duplicated test which is unlikely to fail with anything other than integration issues.

Differences in components perhaps, but yeah, integration issues.  But given that rooting out the cause of such will likely fall on me, I want to know about it ASAP, not a week (or even day) after it happens.  The longer it goes unchecked, the more difficult finding the cause is and I already have way too much to do without adding unnecessary complexity to my tasks without any benefit.

Reply:
We don't want to fail PRs or delay developers for integration issues, what you describe is a good reason for running this on landings.

This stage looks to be about two hours of wall-clock time (in parallel with other stages), of which 90 minutes is running tests, all of which will also be run on CentOS.

If we do want to keep testing PRs on leap then it should be a cut-down set of core tests, not the whole suite (which coincidentally) is what the rest of this PR is about.

Reply:
> We don't want to fail PRs or delay developers for integration issues,

But we also don't want to allow PRs to introduce them.  If PRs are blocked from introducing an integration issue (something that acts differently on a version of a library in EL than it does on Leap for example, perhaps because we are write code that only works on a buggy version of something in EL that then doesn't work on a non-buggy version in Leap) then people shouldn't be subsequently blocked on their PRs.

However if we don't detect and block it at the PR then it lands in master, and is only going to be discovered later that night, (or week perhaps), people then rebasing on that master before it's discovered will be blocked and then it becomes a fire to unblock them.

> what you describe is a good reason for running this on landings.

I would say that what I describe above is a great reason to run it on PRs also.  Or maybe _only_ on PRs if we were really trying to reduce the amount of VM cluster testing.  But we don't actually need to do that.  VM resources are plentiful and cheap.

> This stage looks to be about two hours of wall-clock time (in parallel with other stages), of which 90 minutes is running tests, all of which will also be run on CentOS.

On the most recent [master with both VM stages complete, ](https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/master/3306/pipeline) there was a 20s difference (out of 91 minutes) in the two stages main test step.  The only other meaningful difference in execution time was an extra 11 minutes to run the Leap 15 post provision step.

My guess at that would be that Leap needs to upgrade more packages at the end to get it fully up-to-date.  The Leap update does include a kernel, which means a dracut run, that EL7 doesn't do.  It could be that the Leap base images are further out (i.e. have not been regenerated as recently) of date as the EL7 images.

If that extra 10 minutes on resources that are not constrained is really an issue, my suggestion would be to eliminate it rather than eliminating a test platform that ultimately costs us nothing in time-to-result but does cost us in engineering time when it's elimination allows regressions to be introduced to master.

> If we do want to keep testing PRs on leap then it should be a cut-down set of core tests, not the whole suite (which coincidentally) is what the rest of this PR is about.

Perhaps.  But why?  VM testing is not a pain point for us currently.  I would argue that even with all of the sleeps, it's really our shortest pole.  Looking at the queue right now, the longest waiting job is 18h.  The longest wait for a VM cluster is 11h.

Reply:
Then how about disabling functional on Centos rather than functional on leap?  That gets us the leap testing without duplicating, and we already have the hardware tests running on CentOS.

We do need to get on top of CI runtimes, it's just impractical and simply not working attempting to do the amount of testing that we're currently aiming for.

This PR is nowhere near production ready, it's more about working out what's possible, and what practical changes we could make to reduce the burden.  I absolutely agree that we should have a discussion and make positive decisions about what we should be doing and where the tradeoffs are, but this PR is about gathering data so we can have an informed discussion in a different forum.